VSRAD.DebugServer/Server.cs

```diff
@@ -49,19 +61,77 @@ public async Task LoopAsync()
             while (true)
             {
                 var client = await _listener.AcceptTcpClientAsync().ConfigureAwait(false);
-                ClientConnected(client, clientsCount);
+                TcpClientConnected(client, clientsCount);
                 clientsCount++;
             }
         }
 
-        private void ClientConnected(TcpClient tcpClient, uint clientId)
+        private void TcpClientConnected(TcpClient tcpClient, uint clientId)
         {
             var networkClient = new NetworkClient(tcpClient, clientId);
             var clientLog = new ClientLogger(clientId, _verboseLogging);
+            if (!Task.Run(() => TryProcessServerHandshake(tcpClient, clientLog)).Result)
+            {
+                clientLog.HandshakeFailed(networkClient.EndPoint);
+                return;
+            }
             clientLog.ConnectionEstablished(networkClient.EndPoint);
             Task.Run(() => BeginClientLoopAsync(networkClient, clientLog));
         }
 
+        private async Task<bool> TryProcessServerHandshake(TcpClient client, ClientLogger clientLog)
+        {
+            try
+            {
+                StreamWriter writer = new StreamWriter(client.GetStream(), Encoding.UTF8) { AutoFlush = true };
+                StreamReader reader = new StreamReader(client.GetStream(), Encoding.UTF8);
+
+                // Send server version to client
+                //
+                await writer.WriteLineAsync(_serverVersion.ToString()).ConfigureAwait(false);
+
+                // Obtain client version
+                //
+                String clientResponse = await reader.ReadLineAsync().ConfigureAwait(false);
+
+                Version clientVersion = null;
+                if (!Version.TryParse(clientResponse, out clientVersion))
```

Comment:
Also, to better match project codestyle (and my personal preference, this is not necessary), I would go with single line here:

```C#
if (!Version.TryParse(clientResponse, out var clientVersion))
```

Reply:
done